experiment/secret-sync-controller/secret-sync-config.go

```diff
@@ -0,0 +1,97 @@
+package main
+
+import (
+	"context"
+	// b64 "encoding/base64"
+	// "bytes"
+	"fmt"
+	"gopkg.in/yaml.v2"
+	"strconv"
+	"strings"
+
+	secretmanager "cloud.google.com/go/secretmanager/apiv1"
+	secretmanagerpb "google.golang.org/genproto/googleapis/cloud/secretmanager/v1"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type SecretSyncConfig struct {
+	Specs []SecretSyncSpec `yaml:"specs"`
+}
+
+type SecretSyncSpec struct {
+	// the target can be either a K8s secret or a SecretManager secret
+	Source      Target `yaml:"source"`
+	Destination Target `yaml:"destination"`
+}
+
+type Target struct {
+	// assert that one of the two should be `nil`
+	Kubernetes    *KubernetesSpec    `yaml:"kubernetes,omitempty"`
+	SecretManager *SecretManagerSpec `yaml:"secretManager,omitempty"`
+}
+
+type KubernetesSpec struct {
+	Namespace string   `yaml:"namespace"`
+	Secret    string   `yaml:"secret,omitempty"`
+	DenyList  []string `yaml:"denyList,omitempty"`
+}
+
+type SecretManagerSpec struct {
+	Project  string   `yaml:"project"`
+	Secret   string   `yaml:"secret,omitempty"`
+	DenyList []string `yaml:"denyList,omitempty"`
+}
+
+func (target Target) GetLatestSecretVersion(k8sClientset *kubernetes.Clientset, secretManagerCtx context.Context, secretManagerClient *secretmanager.Client) (int, map[string][]byte) {
+	if k8s, gsm := target.Kubernetes, target.SecretManager; k8s != nil {
+		return k8s.LatestSecretVersion(k8sClientset)
+	} else {
+		return gsm.LatestSecretVersion(secretManagerCtx, secretManagerClient)
+	}
+}
+
+func (k8s *KubernetesSpec) LatestSecretVersion(k8sClientset *kubernetes.Clientset) (int, map[string][]byte) {
+	secret, err := k8sClientset.CoreV1().Secrets(k8s.Namespace).Get(k8s.Secret, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		fmt.Println(err)
+	}
+	version, _ := strconv.Atoi(secret.ObjectMeta.ResourceVersion)
```

Comment:
It looks like you're going through acrobatics to return version as an int both here and below.  I'm not sure this is a safe conversion (is ReourceVersion always guaranteed to be a string representation of an `int`?).  I would probably just keep them as strings instead.

Reply:
You're right about this.
After looking more into resource versions:
https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions
It turns out that they should be treated as strings.
"For example, clients must not assume resource versions are numeric, and may only compare two resource version for equality"

Reply:
According to this: https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes
I think internally they are numeric, but they should be treated as opaque by clients.